File: cat2/cat2stdio.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/stat.h>
#include <fcntl.h>

static void do_cat(const char *path);
static void die(const char *s);

int main(int argc, char *argv[]){
	int i;

	if (argc < 2) {
		fprintf(stderr, "%s: file name not given\n", argv[0]);
		exit(1);
	}

	for (i = 1; i < argc; i++) {
		do_cat(argv[i]);
	}

	exit(0);
}

#define BUFFER_SIZE 2048

static void do_cat(const char *path) {
	FILE *f;
	unsigned char buf[BUFFER_SIZE];
	int n;

	f = fopen(path, "r");
	if (!f) die(path);
	if (f < 0) die(path);
	for (;;) {
		size_t nRead = fread(buf, 1, sizeof buf, f);
		if (ferror(f)) die(path);
		size_t nWrite = fwrite(buf, 1, nRead, stdout);
		if (nWrite < nRead) die(path);
		if (nRead < sizeof buf) break;
	}
	if (fclose(f) != 0) die(path);
}

static void die(const char *s) {
	perror(s);
	exit(1);
}

File: cat2/cat2stdio.swift
import Foundation
#if os(Linux)
	import Glibc
#endif

var standardError = FileHandle.standardError
extension FileHandle : TextOutputStream {
	public func write(_ string: String) {
		guard let data = string.data(using: .utf8) else { return }
		self.write(data)
	}
}

func die(s: String) {
	perror(s);
	exit(1);
}

func do_cat(path: String) {
	var f: UnsafeMutablePointer<FILE>?
	var buf: [CChar] = []

	f = fopen(path, "r")
	if f == nil { die(s: path) }

	while true {
		var nRead: size_t = fread(&buf, 1, 2048, f)
		if ferror(f) != 0 { die(s: path) }
		var nWrite: size_t = fwrite(buf, 1, nRead, stdout)
		if nWrite < nRead { die(s: path)}
		if nRead > buf.count {break}
	}
	fclose(f)

}

let argc = CommandLine.arguments.count
let argv = CommandLine.arguments

if argc < 2 {
	print("\(argv[0]): file name not given", to: &standardError)
}
for file in argv[1..<argc] {
	do_cat(path: file)
}
